src/log/TcpLogger.cc

```diff
@@ -96,6 +96,12 @@ void
 Log::TcpLogger::swanSong()
 {
     disconnect(); // optional: refcounting should close/delete conn eventually
+    if (reconnectScheduled) {
+        // callers must manually manage event queue entries
+        reconnectScheduled->cancel("Log::TcpLogger::swanSong");
```

Comment:
Please remember your earlier requirement for EventScheduler::remove implementation:
 > This explicit cancellation is not necessary before we delete the event object. I would remove this line, but if you insist on having it here, we should add a source code comment to justify its existence.

The coding practice governing this implementation is:
  "queue should not be playing with the internal behaviour of its callers".

Reply:
> N.B. When interpreted correctly, this change request does not contradict [#892 (comment)](https://github.com/squid-cache/squid/pull/892#discussion_r698056369)

There is a contradiction here because callers are not always removing due to cancellation. Sometimes they are removing due to being about to schedule the Call themselves, or just to prevent the Call being scheduled (if possible) while some conflicting activity happens.

Reply:
> > N.B. When interpreted correctly, this change request does not contradict [#892 (comment)](https://github.com/squid-cache/squid/pull/892#discussion_r698056369)

> There is a contradiction here 

We do not need to agree on whether there is a contradiction. If you wish, I can address this change request myself without contradicting that earlier change request. If you have any specific questions that may help your address this change request instead, please ask.

> callers are not always removing due to cancellation

I assert that all existing and any reasonable CancelEvent(call) callers want that call canceled (because they do not want to receive that call -- that is the purpose of that function!). If you can give any specific examples that contradict this assertion, please do. Otherwise, please implement the necessary changes or ask me to implement them instead.

> Sometimes they are removing due to being about to schedule the Call themselves,

That would be a bug. The call may be in the AsyncCallQueue already and one cannot fire the same AsyncCall object twice. Fortunately, there should be no reason to do what you are describing.

> or just to prevent the Call being scheduled (if possible) while some conflicting activity happens.

It is always possible to cancel these events. That guarantee is one of the primary reasons we are adding this new API. And this valid use case clearly belongs my assertion that such an CancelEvent(call) caller will want to cancel the previously scheduled event/call.

Reply:
> > > N.B. When interpreted correctly, this change request does not contradict [#892 (comment)](https://github.com/squid-cache/squid/pull/892#discussion_r698056369)
> 
> > There is a contradiction here
> 
> We do not need to agree on whether there is a contradiction. If you wish, I can address this change request myself without contradicting that earlier change request. If you have any specific questions that may help your address this change request instead, please ask.
> 
> > callers are not always removing due to cancellation
> 
> I assert that all existing and any reasonable CancelEvent(call) callers want that call canceled (because they do not want to receive that call -- that is the purpose of that function!). If you can give any specific examples that contradict this assertion, please do. Otherwise, please implement the necessary changes or ask me to implement them instead.
> 

Which leaves a scheduled entry to a dead Call...


> > Sometimes they are removing due to being about to schedule the Call themselves,
> 
> That would be a bug. The call may be in the AsyncCallQueue already and one cannot fire the same AsyncCall object twice. Fortunately, there should be no reason to do what you are describing.
> 

... you earlier *required* [ removal of unnecessary Calls from the queue.](https://github.com/squid-cache/squid/pull/892#discussion_r700241043)


> > or just to prevent the Call being scheduled (if possible) while some conflicting activity happens.
> 
> It is always possible to cancel these events. That guarantee is one of the primary reasons we are adding this new API. > And this valid use case clearly belongs my assertion that such an CancelEvent(call) caller will want to cancel the previously scheduled event/call.

So ...,  layering C-style wrappers to perform Haskell/Lisp style functional programming API to a C++ program ... in the name of hiding code syntax which C++ developers will be familiar with enough to understand without any such wrappers.  The API design you are trying to force is unnecessary bloat.

Reply:
> Which leaves a scheduled entry to a dead Call...

I am not sure what you mean, but leaving a canceled call object in the global AsyncCallQueue is not a problem. IIRC, I have already explained why. Please let me know if I should spend time to find a reference to that discussion.

> The API design you are trying to force is unnecessary bloat.

Addressing this change request does not lead to bloat. It actually _decreases_ code duplication and improves code reliability.

----

>  layering C-style wrappers to perform Haskell/Lisp style functional programming API to a C++ program

I also recommend dropping "this is C/Haskell/Lisp-style/syntax" arguments. All of the requested changes use regular C++ programming style/syntax/approach. Just because a piece of code looks similar to what could have been written in another language, does not mean it is not a good C++ code. Good C++ programs use stand-alone functions (among other things), just like good programs in many other languages do.